### PR TITLE
refactor: integrate fiddle-core

### PR DESCRIPTION
**Pull request by aryanshridhar**
Integrates fiddle core into the `renderer` codebase

<strong>Todos: </strong>

* ~Fix the `TODOS` over the codebase. (Only one left, not a major one!)~
* ~Fix jests tests (to be done later)~
* Handle `bisecting` of versions through fiddle-core.
* ~Handle different source mirrors in fiddle-core (`global`/`china`)~
* ~Add the `downloadProgress` callback to `fiddle-core` (Opened https://github.com/electron/fiddle-core/pull/30 for the same)~

Tackles #644 as well :)

### CODE REVIEW

File: src/interfaces.ts
Comment (VerteDinde) [inline src/interfaces.ts:10]: Why remove the enum here? Is there a way we can retain this type, so we can make sure that only allowed strings are used when working with versions?

Comment (aryanshridhar) [reply src/interfaces.ts:10]: Oh yeah! But rather than creating a new enum here, it's better if we use the [one defined](https://github.com/electron/fiddle-core/blob/9007a90720c98e46e749b241cfe3d6bbf4d374c3/src/installer.ts#L22) in fiddle-core itself, right ?

Comment (aryanshridhar) [reply src/interfaces.ts:10]: Oh yeah! Probably, the same should be done in fiddle-core as well right ?
[(one such example)](https://github.com/electron/fiddle-core/blob/9007a90720c98e46e749b241cfe3d6bbf4d374c3/src/installer.ts#L180)

File: src/renderer/runner.ts
Comment (aryanshridhar) [inline src/renderer/runner.ts:380]: While fixing tests -- realized that this function always returns `RunResult.INVALID`
Fixed this now :)

File: tests/renderer/state-spec.ts
Comment (aryanshridhar) [inline tests/renderer/state-spec.ts:80]: This timeout waits for the `appState.runner` promise to resolve (most likely!) Otherwise CI throws an error saying -
```
ReferenceError: You are trying to `import` a file after the Jest environment has been torn down. From tests/renderer/state-spec.ts.
```
(As replicated [here](https://github.com/electron/fiddle/runs/7584809546?check_suite_focus=true#step:7:10))

I am quite baffled by this as I thought something like - `await appState.runner` will wait for the promise. Tried that but still CI throws an error.
After many rounds of trial and error, I came up with this _ugly_ solution. Let me know on how can this be resolved in a much better way(If not, then probably - I will add a comment to this code)

File: src/renderer/state.ts
Comment (aryanshridhar) [inline src/renderer/state.ts:178]: This `runner` variable can be confused as an instance of [`Runner`](https://github.com/electron/fiddle/blob/440a94b4f833b077bf576022960d4fe737f04382/src/renderer/runner.ts#L30) class. Open to suggestions for better naming for this variable :) (Probably, `coreRunner`?)

**Comment by aryanshridhar** [comment]
Added a couple of comments above that I felt can be improved. Would love opinions upon it :)

File: src/renderer/state.ts
Comment (dsanders11) [inline src/renderer/state.ts:600]: Is this comment meant to be a TODO? I don't see the context here.

File: src/renderer/components/dialog-add-version.tsx
Comment (dsanders11) [inline src/renderer/components/dialog-add-version.tsx:12]: Since these are already in alphabetical order, let's keep them that way and move this up to be in alphabetical order.

File: src/renderer/components/settings-electron.tsx
Comment (dsanders11) [inline src/renderer/components/settings-electron.tsx:324]:
```suggestion
        // The only way for a local version to be missing
```

File: src/renderer/transforms/forge.ts
Comment (dsanders11) [inline src/renderer/transforms/forge.ts:4]: Let's keep this in the grouping at the top with `child_process`.

**Review by dsanders11 — COMMENTED** [review]
I've done a first pass at review but haven't gone that deep yet. I've left inline comments, and here's a couple top-level comments:

* I'm seeing `yarn.lock` change when running `yarn install`, I think there's a couple dependencies that you missed in the updating of that file in this PR
* If I click "Download" on a version under "Settings", when I close "Settings" the version run button at the top of Fiddle is showing "Checking Status" and spins indefinitely, until I choose a different version from the selector.

File: src/interfaces.ts
Comment (VerteDinde) [reply src/interfaces.ts:10]: I agree with this - the strings in the enum should match their string values

File: src/renderer/runner.ts
Comment (VerteDinde) [inline src/renderer/runner.ts:374]: Any reason why this was removed, or was this re-added in a different area? We do want to have the ability to inject cli flags here if we can

File: src/renderer/state.ts
Comment (VerteDinde) [inline src/renderer/state.ts:170]: Calling out the TODO here: what do we not want to use in production? Is there anything we need to change here before we could ship this in Fiddle?

File: src/renderer/state.ts
Comment (VerteDinde) [reply src/renderer/state.ts:178]: Depending on the functionality of the other existing class, something like `versionRunner` may work

File: src/renderer/state.ts
Comment (VerteDinde) [inline src/renderer/state.ts:302]: Nit:
```suggestion
      // A local electron build's `state` is setup in versions.ts
```

File: src/renderer/state.ts
Comment (VerteDinde) [inline src/renderer/state.ts:559]:
```suggestion
        if (this.installer.state(version) === VersionState.missing) {
```

File: src/renderer/state.ts
Comment (VerteDinde) [inline src/renderer/state.ts:920]: Is this dependent on a new release of fiddle-core, or does this still need to be changed in fiddle-core?

File: src/renderer/state.ts
Comment (VerteDinde) [inline src/renderer/state.ts:937]: Same comment as David had above, let's get these string values and key aligned.

**Review by VerteDinde — CHANGES_REQUESTED** [review]
Rough first pass review, left a few comments.

File: src/renderer/state.ts
Comment (aryanshridhar) [reply src/renderer/state.ts:170]: This is because `BaseVersion` is [documented](https://github.com/electron/fiddle-core/blob/main/src/versions.ts#L84) to be a only used while testing.
Our use case here - If fiddle-core is not populated with versions, it tries to fetches the version from [here](https://github.com/electron/fiddle-core/blob/main/src/versions.ts#L203). This causes a couple of errors - 
* We already have this code and electron version are fetched in fiddle prior.
* Since, it uses [node-fetch](https://github.com/electron/fiddle-core/blob/main/src/versions.ts#L4), it throws an error when this function is called here. (The error that I mentioned [here](https://electronhq.slack.com/archives/C03G7RWF6KH/p1657741854218999))

File: src/renderer/components/version-select.tsx
Comment (aryanshridhar) [reply src/renderer/components/version-select.tsx:106]: Oh yup, right! Thanks :)

File: src/interfaces.ts
Comment (aryanshridhar) [reply src/interfaces.ts:10]: Oh Cool! I felt it will be more intuitive for the user. (As `installing` is not much intuitive signifying that the version is actually being getting unzipped)

File: src/renderer/state.ts
Comment (aryanshridhar) [reply src/renderer/state.ts:600]: Oh yeah! This is to be removed :)

File: src/renderer/components/dialog-add-version.tsx
Comment (aryanshridhar) [reply src/renderer/components/dialog-add-version.tsx:12]: How about we introduce the [sort-import ](https://eslint.org/docs/latest/rules/sort-imports)rule in eslint rather than doing it manually everytime ?

File: src/renderer/transforms/forge.ts
Comment (aryanshridhar) [reply src/renderer/transforms/forge.ts:4]: Same comment as mentioned here.
Maybe its better if I put up a thread on slack regarding it :)

File: src/renderer/state.ts
Comment (aryanshridhar) [reply src/renderer/state.ts:920]: I added this because it this can be integrated in fiddle-core and then imported here.
(We are passing the states as string in fiddle-core, [one such example](https://github.com/electron/fiddle-core/blob/e22ae4d453ba21c5d40f47b153105f632a624c9a/src/installer.ts#L117)).

Comment (aryanshridhar) [reply src/renderer/state.ts:920]: Local branch changes - (https://github.com/electron/fiddle-core/compare/main...aryanshridhar:fiddle-core:Add-VersionStateEnum?expand=1)

Comment (aryanshridhar) [reply src/renderer/state.ts:920]: I have opened up a pr for it - https://github.com/electron/fiddle-core/pull/33
Once this is merged, we don't need to have this extra check.

File: src/renderer/runner.ts
Comment (aryanshridhar) [reply src/renderer/runner.ts]: Yup, nice catch! Missed that!

**Comment by aryanshridhar** [comment]
Updates on the pr - 

> If I click "Download" on a version under "Settings", when I close "Settings" the version run button at the top of Fiddle is showing "Checking Status" and spins indefinitely, until I choose a different version from the selector.

* Debugged this and turns out fiddle-core blocks the installation of other versions if one is already installing (https://github.com/electron/fiddle-core/pull/35)
* Rebased the pr so now the imports are always sorted :)
* Removed the VersionState interface to avoid deduplication of code (https://github.com/electron/fiddle-core/pull/33). We can now directly import the interface from fiddle-core and use it.

Also, would love reviews on [this](https://github.com/electron/fiddle/pull/1134#discussion_r935013423) comment!

**Comment by aryanshridhar** [comment]
Rebased and now the checks are passing :rocket: (A pr to fiddle-core upgrading `get` is only left out)

File: src/renderer/state.ts
Comment (erickzhao) [reply src/renderer/state.ts:168]: Resolving as we said it was okay in our weekly mentor meeting. 👍

File: src/interfaces.ts
Comment (VerteDinde) [inline src/interfaces.ts:1]: One merge-blocker - testing with a fork of `fiddle-core` is perfectly fine, but we need to cut a new release and use the real `fiddle-core` package before we merge this into Fiddle's main branch.

Comment (aryanshridhar) [reply src/interfaces.ts:1]: Yup right! The main branch of `electron/fiddle-core` has all the right set of APIs required. We just need to release it :)

File: src/renderer/runner.ts
Comment (erickzhao) [inline src/renderer/runner.ts:199]: question: It's been a while since I've taken a look at `fiddle-core` code. What is the difference between `InstallState.installed` and `InstallState.downloaded` if they are both considered ready to run here?

**Review by erickzhao — CHANGES_REQUESTED** [review]
First pass with a single blocking change request

File: src/renderer/runner.ts
Comment (aryanshridhar) [reply src/renderer/runner.ts:199]: `downloaded` is the state when the binary is downloaded and kept as a `zip` file whereas the `installed` is when the downloaded binary is currently unzipped for the fiddle to run (`fiddle-core` doesn't unzip all downloaded versions to improve disk space usage)
In our use case - We want both the states to be marked `ready` since `fiddle-core.spawn` by default unzips the version if it's not installed.

File: src/renderer/runner.ts
Comment (aryanshridhar) [reply src/renderer/runner.ts:200]: Nope, I can't think of a such case where they will be different.
This is because - everytime the version is changed, it triggers [this](https://github.com/aryanshridhar/fiddle/blob/Integrate-fiddlecore/src/renderer/state.ts#L373-L376) event which updates the state of `currentElectronVersion`.

File: src/interfaces.ts
Comment (aryanshridhar) [reply src/interfaces.ts:9]: Yeah, What `fiddle` does is similar to what `fiddle-core` does.
We mark versions to be `missing` if they are either [removed](https://github.com/electron/fiddle-core/blob/7d17028513f853abbe5eb3a6ebf99d326252e1ef/src/installer.ts#L190-L192) by user or not being currently [`downloaded/installed`](https://github.com/electron/fiddle-core/blob/main/src/installer.ts#L82-L84).

File: src/renderer/state.ts
Comment (aryanshridhar) [reply src/renderer/state.ts:608]: Debugged this and turns out that the issue is way deep! Probably with the new `get` version.
(I will open a new thread regarding it in slack)

File: src/renderer/constants.ts
Comment (aryanshridhar) [reply src/renderer/constants.ts:11]: Oh yeah! However, we won't be still able to use the existing downloaded versions as `fiddle-core` keeps them in a zip file and also uses a different naming convention of files.

Comment (erickzhao) [reply src/renderer/constants.ts:11]: That's a big breaking change for users (especially if they have 10+ Electron versions downloaded), and I don't think I'm super comfortable releasing this as-is.

Is there a way we can either:
1. Run a startup script that moves Fiddle scripts to a format that fiddle-core understands?
2. Make changes to Fiddle Core so it understands folders rather than zips?

Comment (VerteDinde) [reply src/renderer/constants.ts:11]: I agree with Erick, this is a pretty major change for existing Fiddle users. Of the two options, could we modify fiddle-core to understand both zip and folder structures? I feel like that would be the more future-proof option

Comment (BlackHole1) [reply src/renderer/constants.ts:11]: I also agree with the second.
The first one will be more costly, if the user downloads very many versions of the electron, the migration script will be very slow to execute and will predictably have a lot of error handling logic. And with a lot of unit tests to go along with it

Comment (BlackHole1) [reply src/renderer/constants.ts:11]: One more thing to consider: when using fiddle to load a locally built electron, the electron is a folder, not a zip file.

image

**Review by VerteDinde — APPROVED** [review]
Great work! Approved, contingent on changing the fiddle-core version from the fork to the main package 👍

**Review by erickzhao — APPROVED** [review]

**Comment by ckerr** [comment]
So cool that this is in `main` now. Thank you @aryanshridhar, this is great! :heart:

**Pull request merged**
